src/persist-client/src/read.rs

```diff
@@ -523,18 +687,49 @@ where
         trace!("ReadHandle::clone");
         let new_reader_id = ReaderId::new();
         let mut machine = self.machine.clone();
-        let read_cap = machine.clone_reader(&self.reader_id).await;
+        let read_cap = machine.clone_reader(&new_reader_id, (self.cfg.now)()).await;
         let new_reader = ReadHandle {
+            cfg: self.cfg.clone(),
             metrics: Arc::clone(&self.metrics),
             reader_id: new_reader_id,
             machine,
             blob: Arc::clone(&self.blob),
             since: read_cap.since,
+            last_heartbeat: Instant::now(),
             explicitly_expired: false,
         };
         new_reader
     }
 
+    /// If necessary, heartbeats the read lease via [Self::downgrade_since].
+    ///
+    /// This is an internally rate limited helper, designed to allow users to
+    /// call it as frequently as they like. Call either this or
+    /// [Self::downgrade_since] on some interval that is "frequent" compared to
+    /// PersistConfig::FAKE_READ_LEASE_DURATION.
+    ///
+    /// If you'd like to rate-limit progress information, feel free to provide a
+    /// real since here and use this instead (or in addition to)
+    /// downgrade_since. This is given preferential treatment over non-progress
+    /// heartbeats.
+    pub async fn maybe_heartbeat(&mut self, new_since: Option<&Antichain<T>>) {
```

Comment:
I find the name of this function confusing, since it's sometimes a `downgrade_since` call (with a heartbeat attached), and sometimes just a heartbeat. Since snapshots only call this without `new_since`, and listens only call it with `new_since`, maybe it should be two different methods?

Reply:
yeah! I think that'd help clarify intent